src/kv.c

```diff
@@ -1223,7 +1248,13 @@ kvStore_Keys(kvKeysList *list, kvStore *kv, kvWatchOptions *opts)
         h = h->next;
         if (s == NATS_OK)
         {
-            DUP_STRING(s, list->Keys[i], e->key);
+            if(filters!=NULL && numFilters > 0) {
```

Comment:
Same, I don't understand why you have to treat this differently than a WatchAll(). If there is a filter and the server is doing proper filtering, you will get anything you need same than WatchAll().

The point of my comment of making it this way was because I assumed that it was just a matter of passing filters vs the single filter ">". But rest was the same. I checked the Java implementation and does not look like he changed anything.

Reply:
In java we have the following:
https://github.com/nats-io/nats.java/blob/d69587382c854abe7c4d840cb85c5846e7b0351e/src/main/java/io/nats/client/impl/NatsKeyValue.java#L305
There is an explicit filtering for PUT kv operation type.

My assumption is if the server is not returning deleted/purged messages, then the IgnoreDelete option set as true should be indicating that. In that case why is the PUT operation explicitly handled in java client.

Reply:
We set the IgnoreDelete set to true, which means that internally, when we get the result from the server, we will ignore deletes and already collect only "PUT" operations. So you don't need to do any additional work (see my diff I sent, it works with your new test):

https://github.com/nats-io/nats.c/blob/381f01cb75acb6ff0860a599ff29b17127d21dac/src/kv.c#L1005

Reply:
In other words, where we "dup" the strings, we have already filtered properly in `KV_GATHER_LIST;`

Reply:
This was based on the go client: https://github.com/nats-io/nats.go/blob/b61c7c554f188ff40e15fb3c4668bad959ccde86/kv.go#L1012

Reply:
Thanks for the explanation. I went through the go client as well. I was redoing the work already done in kvWatcher_Next.